File: src/models/postModel.js
const mongoose = require('mongoose');
const slugify = require('slugify');

const postSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'A post must have a title'],
    trim: true,
    maxlength: [100, 'A post title cannot be more than 100 characters']
  },
  slug: {
    type: String,
    unique: true
  },
  content: {
    type: String,
    required: [true, 'A post must have content']
  },
  excerpt: {
    type: String,
    maxlength: [200, 'An excerpt cannot be more than 200 characters']
  },
  featuredImage: {
    type: mongoose.Schema.ObjectId,
    ref: 'Media'
  },
  author: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: [true, 'A post must have an author']
  },
  categories: [{
    type: mongoose.Schema.ObjectId,
    ref: 'Category'
  }],
  tags: [String],
  status: {
    type: String,
    enum: ['draft', 'published', 'scheduled', 'archived'],
    default: 'draft'
  },
  publishedAt: Date,
  scheduledFor: Date,
  featured: {
    type: Boolean,
    default: false
  },
  trending: {
    type: Boolean,
    default: false
  },
  readTime: Number,
  viewCount: {
    type: Number,
    default: 0
  },
  likes: {
    type: Number,
    default: 0
  },
  likedBy: [{
    type: mongoose.Schema.ObjectId,
    ref: 'User'
  }],
  metadata: {
    seoTitle: String,
    seoDescription: String,
    keywords: [String],
    canonical: String
  },
  revisions: [{
    content: String,
    updatedAt: Date,
    updatedBy: {
      type: mongoose.Schema.ObjectId,
      ref: 'User'
    }
  }],
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// INDEXES
postSchema.index({ slug: 1 });
postSchema.index({ title: 'text', content: 'text', tags: 'text' });
postSchema.index({ status: 1, scheduledFor: 1 });
postSchema.index({ author: 1 });
postSchema.index({ categories: 1 });
postSchema.index({ featured: 1 });
postSchema.index({ trending: 1 });
postSchema.index({ viewCount: -1 });

// VIRTUAL POPULATE
postSchema.virtual('comments', {
  ref: 'Comment',
  foreignField: 'post',
  localField: '_id'
});

// DOCUMENT MIDDLEWARE: runs before .save() and .create()
postSchema.pre('save', function(next) {
  // Generate slug from title
  if (this.isNew || this.isModified('title')) {
    this.slug = slugify(this.title, { lower: true }) + 
      '-' + 
      Math.floor(1000 + Math.random() * 9000).toString();
  }

  // Generate excerpt if not provided
  if (!this.excerpt && this.content) {
    this.excerpt = this.content
      .replace(/<\/?[^>]+(>|$)/g, '') // Remove HTML tags
      .substring(0, 197) + '...';
  }

  // Calculate read time (average reading speed: 200 words per minute)
  if (this.content) {
    const wordCount = this.content
      .replace(/<\/?[^>]+(>|$)/g, '') // Remove HTML tags
      .split(/\s+/).length;
    this.readTime = Math.ceil(wordCount / 200);
  }

  // Store revision on update
  if (!this.isNew && this.isModified('content')) {
    if (!this.revisions) this.revisions = [];
    this.revisions.push({
      content: this.content,
      updatedAt: Date.now(),
      updatedBy: this.updatedBy // This should be set before saving
    });
  }

  // Update status based on scheduling
  if (
    this.status === 'scheduled' && 
    this.scheduledFor && 
    this.scheduledFor <= new Date()
  ) {
    this.status = 'published';
    this.publishedAt = this.scheduledFor;
    this.scheduledFor = undefined;
  }

  // Set publishedAt date if publishing
  if (
    this.status === 'published' && 
    (!this.publishedAt || this.isModified('status'))
  ) {
    this.publishedAt = Date.now();
  }

  next();
});

// Set updatedAt on update
postSchema.pre('findOneAndUpdate', function(next) {
  this.set({ updatedAt: Date.now() });
  next();
});

// QUERY MIDDLEWARE
postSchema.pre(/^find/, function(next) {
  // Populate references
  this.populate({
    path: 'author',
    select: 'name username photo'
  }).populate({
    path: 'categories',
    select: 'name slug'
  }).populate({
    path: 'featuredImage',
    select: 'url alt'
  });

  next();
});

const Post = mongoose.model('Post', postSchema);

module.exports = Post;

File: flutter_app/lib/data/models/post_model.dart
import '../../domain/entities/post.dart';
import 'user_model.dart';
import 'category_model.dart';

class PostModel extends Post {
  const PostModel({
    required super.id,
    required super.title,
    required super.slug,
    required super.content,
    super.excerpt,
    super.featuredImage,
    required super.author,
    required super.categories,
    required super.tags,
    required super.status,
    required super.isFeatured,
    required super.viewCount,
    required super.likesCount,
    required super.commentsCount,
    required super.publishedAt,
    required super.createdAt,
    required super.updatedAt,
  });

  factory PostModel.fromJson(Map<String, dynamic> json) {
    return PostModel(
      id: json['_id'] ?? json['id'],
      title: json['title'],
      slug: json['slug'],
      content: json['content'],
      excerpt: json['excerpt'],
      featuredImage: json['featuredImage'],
      author: UserModel.fromJson(json['author']),
      categories: (json['categories'] as List)
          .map((category) => CategoryModel.fromJson(category))
          .toList(),
      tags: List<String>.from(json['tags'] ?? []),
      status: json['status'],
      isFeatured: json['isFeatured'] ?? false,
      viewCount: json['viewCount'] ?? 0,
      likesCount: json['likesCount'] ?? 0,
      commentsCount: json['commentsCount'] ?? 0,
      publishedAt: DateTime.parse(json['publishedAt']),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'slug': slug,
      'content': content,
      'excerpt': excerpt,
      'featuredImage': featuredImage,
      'author': (author as UserModel).toJson(),
      'categories': categories.map((category) => (category as CategoryModel).toJson()).toList(),
      'tags': tags,
      'status': status,
      'isFeatured': isFeatured,
      'viewCount': viewCount,
      'likesCount': likesCount,
      'commentsCount': commentsCount,
      'publishedAt': publishedAt.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  PostModel copyWith({
    String? id,
    String? title,
    String? slug,
    String? content,
    String? excerpt,
    String? featuredImage,
    UserModel? author,
    List<CategoryModel>? categories,
    List<String>? tags,
    String? status,
    bool? isFeatured,
    int? viewCount,
    int? likesCount,
    int? commentsCount,
    DateTime? publishedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return PostModel(
      id: id ?? this.id,
      title: title ?? this.title,
      slug: slug ?? this.slug,
      content: content ?? this.content,
      excerpt: excerpt ?? this.excerpt,
      featuredImage: featuredImage ?? this.featuredImage,
      author: author ?? this.author,
      categories: categories ?? this.categories,
      tags: tags ?? this.tags,
      status: status ?? this.status,
      isFeatured: isFeatured ?? this.isFeatured,
      viewCount: viewCount ?? this.viewCount,
      likesCount: likesCount ?? this.likesCount,
      commentsCount: commentsCount ?? this.commentsCount,
      publishedAt: publishedAt ?? this.publishedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}